bpfman-operator/apis/v1alpha1/shared_types.go

```diff
@@ -0,0 +1,383 @@
+	ShouldAttach bool `json:"shouldAttach"`
+	// uuid is an Unique identifier for the attach point assigned by bpfman agent.
+	UUID string `json:"uuid"`
+	// linkId is an identifier for the link assigned by bpfman. This field is
+	// empty until the program is successfully attached and bpfman returns the
+	// id.
+	LinkId *uint32 `json:"linkId"`
+	// linkStatus reflects whether the attachment has been reconciled
+	// successfully, and if not, why.
+	LinkStatus LinkStatus `json:"linkStatus"`
+}
+
+type BpfProgramStateCommon struct {
+	// name is the name of the function that is the entry point for the BPF
+	// program
+	Name string `json:"name"`
+	// programLinkStatus reflects whether all links requested for the program
+	// are in the correct state.
+	ProgramLinkStatus ProgramLinkStatus `json:"programLinkStatus"`
+	// programId is the id of the program in the kernel.  Not set until the
+	// program is loaded.
+	// +optional
+	ProgramId *uint32 `json:"programId"`
+}
+
+// PullPolicy describes a policy for if/when to pull a container image
+// +kubebuilder:validation:Enum=Always;Never;IfNotPresent
+type PullPolicy string
+
+const (
+	// PullAlways means that bpfman always attempts to pull the latest bytecode image. Container will fail If the pull fails.
+	PullAlways PullPolicy = "Always"
+	// PullNever means that bpfman never pulls an image, but only uses a local image. Container will fail if the image isn't present
+	PullNever PullPolicy = "Never"
+	// PullIfNotPresent means that bpfman pulls if the image isn't present on disk. Container will fail if the image isn't present and the pull fails.
+	PullIfNotPresent PullPolicy = "IfNotPresent"
+)
+
+// ByteCodeSelector defines the various ways to reference bpf bytecode objects.
+type ByteCodeSelector struct {
+	// image used to specify a bytecode container image.
+	Image *ByteCodeImage `json:"image,omitempty"`
+
+	// path is used to specify a bytecode object via filepath.
+	Path *string `json:"path,omitempty"`
+}
+
+// ByteCodeImage defines how to specify a bytecode container image.
+type ByteCodeImage struct {
+	// url is a valid container image URL used to reference a remote bytecode image.
+	Url string `json:"url"`
+
+	// pullPolicy describes a policy for if/when to pull a bytecode image. Defaults to IfNotPresent.
+	// +kubebuilder:default:=IfNotPresent
+	// +optional
+	ImagePullPolicy PullPolicy `json:"imagePullPolicy"`
+
+	// imagePullSecret is the name of the secret bpfman should use to get remote image
+	// repository secrets.
+	// +optional
+	ImagePullSecret *ImagePullSecretSelector `json:"imagePullSecret,omitempty"`
+}
+
+// ImagePullSecretSelector defines the name and namespace of an image pull secret.
+type ImagePullSecretSelector struct {
+	// name of the secret which contains the credentials to access the image repository.
+	Name string `json:"name"`
+
+	// namespace of the secret which contains the credentials to access the image repository.
+	Namespace string `json:"namespace"`
```

Comment:
See https://kubernetes.io/docs/concepts/overview/working-with-objects/names for Kubernetes object name validations you can put in place to ensure they are a valid format at admission time